### PR TITLE
Sandbox/si/main/johnson

### PR DESCRIPTION
**Pull request by si-23**

### CODE REVIEW

File: py_import_cycles/modules.py [hunk 1]
```diff
@@ -53,22 +64,51 @@ def joinname(self, *names: str | ModuleName) -> ModuleName:
...
+
+    def __lt__(self, other: object) -> bool:
+        if not isinstance(other, Module):
+            return NotImplemented
+        return self.name < other.name
+
+    def __gt__(self, other: object) -> bool:
+        if not isinstance(other, Module):
+            return NotImplemented
+        return self.name > other.name
```
Comment (Synss) [inline py_import_cycles/modules.py:93]: If you provide these, you should provide all
```
def __ge__(self, other: object) -> bool:
    return self == other or self > other

def __le__(self, other: object) -> bool:
    return self == other or self < other

```
I think `__ne__` is provided automatically but, as far as I know, `>=` and `<=` are not.

File: py_import_cycles/modules.py [hunk 2]
```diff
@@ -53,22 +64,51 @@ def joinname(self, *names: str | ModuleName) -> ModuleName:
...
 
 
-class NamespacePackage(NamedTuple):
-    path: Path
-    name: ModuleName
+class RegularPackage(Module):
+    def _validate(self, path: Path, name: ModuleName) -> None:
+        if path.with_suffix("").name != _INIT_NAME or name.parts[-1] != _INIT_NAME:
```
Comment (Synss) [inline py_import_cycles/modules.py:98]: I don't know whether this applies but can't `name.parts[-1]` raise `IndexError`?

File: py_import_cycles/modules.py [hunk 3]
```diff
@@ -53,22 +64,51 @@ def joinname(self, *names: str | ModuleName) -> ModuleName:
...
+            raise ValueError()
```
Comment (Synss) [inline py_import_cycles/modules.py:99]: maybe better `raise ValueError(path)` and `raise ValueError(name)` (so also split the `if ... or` in two different checks)?

File: py_import_cycles/modules.py [hunk 4]
```diff
@@ -53,22 +64,51 @@ def joinname(self, *names: str | ModuleName) -> ModuleName:
...
 
 
-class PyModule(NamedTuple):
-    path: Path
-    name: ModuleName
+class NamespacePackage(Module):
+    def _validate(self, path: Path, name: ModuleName) -> None:
+        if _INIT_NAME in (path.with_suffix("").name, name.parts[-1]):
+            raise ValueError()
```
Comment (Synss) [inline py_import_cycles/modules.py:105]: same comments as for `RegularPackage`

File: py_import_cycles/modules.py [hunk 5]
```diff
@@ -53,22 +64,51 @@ def joinname(self, *names: str | ModuleName) -> ModuleName:
...
 
 
-Module = RegularPackage | NamespacePackage | PyModule
+class PyModule(Module):
+    def _validate(self, path: Path, name: ModuleName) -> None:
+        if _INIT_NAME in (path.with_suffix("").name, name.parts[-1]):
+            raise ValueError()
```
Comment (Synss) [inline py_import_cycles/modules.py:111]: same comments as for `RegularPackage`

File: Pipfile [hunk 6]
```diff
@@ -11,6 +11,7 @@ mypy = "*"
 
 [packages]
 graphviz = "*"
+networkx = "*"
```
Comment (Synss) [inline Pipfile:14]: probably also in `setup.py`

File: py_import_cycles/modules.py (hunk 2 above)
Comment (si-23) [reply py_import_cycles/modules.py]: done

File: py_import_cycles/modules.py (hunk 1 above)
Comment (si-23) [reply py_import_cycles/modules.py:99]: done

File: py_import_cycles/modules.py (hunk 3 above)
Comment (si-23) [reply py_import_cycles/modules.py]: done

File: py_import_cycles/modules.py (hunk 4 above)
Comment (si-23) [reply py_import_cycles/modules.py]: done

File: py_import_cycles/modules.py (hunk 5 above)
Comment (si-23) [reply py_import_cycles/modules.py]: done

File: Pipfile (hunk 6 above)
Comment (si-23) [reply Pipfile:14]: done

**Pull request merged**